controllers/widget_controller.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	kcpclient "github.com/kcp-dev/apimachinery/pkg/client"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	datav1alpha1 "github.com/kcp-dev/controller-runtime-example/api/v1alpha1"
+)
+
+// WidgetReconciler reconciles a Widget object
+type WidgetReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=data.my.domain,resources=widgets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=data.my.domain,resources=widgets/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=data.my.domain,resources=widgets/finalizers,verbs=update
+
+// Reconcile TODO
+func (r *WidgetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	// Include the clusterName from req.ObjectKey in the logger, similar to the namespace and name keys that are already
+	// there.
+	logger = logger.WithValues("clusterName", req.ObjectKey.Cluster.String())
```

Comment:
Do we expect changes to controller-runtime making this unnecessary in user-code?

Reply:
Maaaaaybe but not in the immediate term